test/integration/cdi/example-driver/app/kubeletplugin.go

```diff
@@ -0,0 +1,151 @@
+}
+
+// newExamplePlugin returns an initialized examplePlugin instance.
+func newExamplePlugin(cdiDir, driverName, draAddress, pluginRegistrationPath string) (*examplePlugin, error) {
+	return &examplePlugin{
+		cdiDir:                 cdiDir,
+		driverName:             driverName,
+		draAddress:             draAddress,
+		pluginRegistrationPath: pluginRegistrationPath,
+	}, nil
+}
+
+// getJsonFilePath returns the absolute path where json file is/should be.
+func (ex *examplePlugin) getJsonFilePath(claimUID string) string {
+	jsonfileName := fmt.Sprintf("%s-%s.json", ex.driverName, claimUID)
+	return filepath.Join(ex.cdiDir, jsonfileName)
+}
+
+// startPlugin starts servers that are necessary for plugins and registers the plugin with kubelet.
+func (ex examplePlugin) startPlugin() error {
+	klog.Infof("Starting kubelet plugin")
+
+	// create CDI directory if not exists
+	if err := os.MkdirAll(ex.cdiDir, os.FileMode(0750)); err != nil {
+		return fmt.Errorf("failed to create directory: %v", err)
+	}
+
+	// create DRA driver socket directory if not exists.
+	if err := os.MkdirAll(filepath.Dir(ex.draAddress), 0750); err != nil {
+		return fmt.Errorf("failed to create DRA driver socket directory: %v", err)
+	}
+
+	// Run a grpc server for the driver, that listens on draAddress
+	if err := plugin.StartNonblockingGrpcServer(ex.draAddress, &ex); err != nil {
+		return fmt.Errorf("failed to start a nonblocking grpc server: %v", err)
+	}
+
+	// Run a registration server and registers plugin with kubelet
+	if err := plugin.StartRegistrar(ex.driverName, ex.draAddress, ex.pluginRegistrationPath); err != nil {
+		return fmt.Errorf("failed to start registrar: %v", err)
+	}
+
+	return nil
+
+}
+
+func (ex *examplePlugin) NodePrepareResource(ctx context.Context, req *drapbv1.NodePrepareResourceRequest) (*drapbv1.NodePrepareResourceResponse, error) {
+
+	kind := ex.driverName + "/" + deviceName
+
+	klog.Infof("NodePrepareResource is called: request: %+v", req)
+
+	klog.Infof("Creating CDI File")
+	// make resourceHandle in the form of map[string]string
+	var decodedResourceHandle map[string]string
+	if err := json.Unmarshal([]byte(req.ResourceHandle), &decodedResourceHandle); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal resourceHandle: %v", err)
+	}
+
+	// create ContainerEdits in device spec and ContainerEdits
+	envs := []string{}
+	for key, val := range decodedResourceHandle {
+		envs = append(envs, key+"="+val)
+	}
+	// create spec
+	spec := specs.Spec{
+		Version: cdiVersion,
+		Kind:    kind,
+		Devices: []specs.Device{{
+			Name: deviceName,
+			ContainerEdits: specs.ContainerEdits{
+				Env: envs,
+			},
+		}},
+		ContainerEdits: specs.ContainerEdits{},
+	}
+	cdiSpec, err := cdiapi.NewSpec(&spec, ex.getJsonFilePath(req.ClaimUid), 1)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new spec: %v", err)
+	}
+	klog.V(5).InfoS("CDI spec is validated")
+
+	// create bytes from spec, which would be written into the json file
```

Comment:
There is a [cdiSpec.Write() API](https://github.com/container-orchestrated-devices/container-device-interface/blob/master/pkg/cdi/spec.go#L116) that does exactly what we need here. I've asked CDI maintainer to tag it. After that we can use it here.